### PR TITLE
add parentwork plugin

### PR DESCRIPTION
**Pull request by dosoe**
New try, closes #2580 .

### CODE REVIEW

File: beetsplug/parentwork.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+from __future__ import division, absolute_import, print_function
+
+from beets import ui
+from beets.plugins import BeetsPlugin
+
+import musicbrainzngs
+
+
+def work_father(mb_workid, work_date=None):
+    """ This function finds the id of the father work given its id"""
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: The naming is a little unclear here—what exactly is a "father work" and how is it different from a "parent work"? Maybe something like "parent" vs. "ancestor" would be clearer? Also, it would be nice if the docstring described (in a sentence or two) exactly what's these things mean instead of just referencing the name.

Also, this function seems to return two things, not just one.

Comment (sampsyo) [reply beetsplug/parentwork.py]: Also, FWIW, it's not necessary to say "This function does X…" It's usually more appropriate to just say "Do X." This series of examples might be helpful:
https://blog.dolphm.com/pep257-good-python-docstrings-by-example/

File: beetsplug/parentwork.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    work_date = None
+    while True:
+        (new_mb_workid, work_date) = work_father(mb_workid, work_date)
+        if not new_mb_workid:
+            return mb_workid, work_date
+        mb_workid = new_mb_workid
+    return mb_workid, work_date
+
+
+def find_parentwork(mb_workid):
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: Similarly, with the naming: `find_parentwork` sounds like it does the same thing as `work_parent`. :smiley: Naming some of the functions to use "id" might help.

File: beetsplug/parentwork.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+class ParentWorkPlugin(BeetsPlugin):
+    def __init__(self):
+        super(ParentWorkPlugin, self).__init__()
+
+        self.config.add({
+            'auto': False,
+            'force': False,
+        })
+
+        self._command = ui.Subcommand(
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: It's probably more appropriate to declare the `ui.Subcommand` inside the `commands` method instead of making it a field. (It's only used there.)

File: beetsplug/parentwork.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        mb_parentworkid = work_info['work']['id']
+        if 'disambiguation' in work_info['work']:
+            parentwork_disambig = work_info['work']['disambiguation']
+            return [parentwork, mb_parentworkid, parent_composer,
+                    parent_composer_sort, parentwork_disambig]
+        else:
+            return [parentwork, mb_parentworkid, parent_composer,
+                    parent_composer_sort, None]
+
+    def find_work(self, item, force):
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: This could use a docstring describing which fields on `item` get updated.

File: beetsplug/parentwork.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+        recording_id = item.mb_trackid
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: I think it would be OK (and a little clearer) to just use `item.mb_trackid` instead of putting it in a variable here.

File: beetsplug/parentwork.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        try:
+            item.parentwork
+            hasparent = True
+        except AttributeError:
+            hasparent = False
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: I recommend just using `hasattr(item, 'parentwork')`.

File: beetsplug/parentwork.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+                found = True
+            except musicbrainzngs.musicbrainz.WebServiceError:
+                self._log.debug("Work unreachable")
+                found = False
+        elif hasparent:
+            self._log.debug("Work already in library, not necessary fetching")
+            return
+
+        if found:
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: I don't think you need `found` or this conditional—just put this right in the condition above after gathering the data. (Or perhaps more appropriately, after an "early return.")

File: beetsplug/parentwork.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                work_info, work_date = find_parentwork(item.mb_workid)
+                parent_info = self.get_info(item, work_info)
+                parentwork = parent_info[0]
+                mb_parentworkid = parent_info[1]
+                parent_composer = parent_info[2]
+                parent_composer_sort = parent_info[3]
+                parentwork_disambig = parent_info[4]
+
+                found = True
+            except musicbrainzngs.musicbrainz.WebServiceError:
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: A `try`/`except` should go only around the code that might raise the appropriate exception. Here, I think that's just `find_parentwork`?

File: beetsplug/parentwork.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        force_parent = self.config['force'].get(bool)
+
+        for item in task.imported_items():
+            self.find_work(item, force_parent)
+            item.store()
+
+    def get_info(self, item, work_info):
+        """Given the parentwork info dict, this function fetches
+        parent_composer, parent_composer_sort, parentwork,
+        parentwork_disambig, mb_workid and composer_ids"""
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: The return type here is a little awkward—perhaps consider returning a `dict` whose fields are exactly what needs to be assigned onto the item?

File: beetsplug/parentwork.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+            item['parentwork'] = parentwork
+            if parentwork_disambig:
+                item['parentwork_disambig'] = parentwork_disambig
+            item['mb_parentworkid'] = mb_parentworkid
+            item['parent_composer'] = u''
+            item['parent_composer'] = u', '.join(parent_composer)
+            item['parent_composer_sort'] = u''
+            item['parent_composer_sort'] = u', '.join(parent_composer_sort)
+            if work_date:
+                item['work_date'] = work_date
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: Maybe there's a simpler way to do all these assignments—something along the lines of:

```py
for key, value in info.items():
    if value:
        item[key] = value
```

…especially if the return type of `get_info` is changed to a dict.

**Comment by dosoe** [comment]
It's very fun, I'm correcting my students' informatics exams (basic python course) and I'm getting informatics teached at the same time. Thanks for the comments! I never got healthy programming practices teached (at least not on this level).

File: beetsplug/parentwork.py (hunk 8 above)
Comment (dosoe) [reply beetsplug/parentwork.py]: Strictly speaking, it is just find_father, but there I cannot use self._log.debug() as it is not in the ParentWorkPlugin class.

File: beetsplug/parentwork.py (hunk 1 above)
Comment (dosoe) [reply beetsplug/parentwork.py]: Thanks for the link! Yes, in the old PR I got requested to fetch the composition date, so I fetch it at each level until I find a date. It's not very satisfactory, I feel like this plugin is not really the place to do this, but it's possible.

File: beetsplug/parentwork.py (hunk 3 above)
Comment (dosoe) [reply beetsplug/parentwork.py]: I have no idea how to do this and how it works, I just copy-pasted it from another plugin. Isn't there a difference between ```_command``` and ```command```?

File: beetsplug/parentwork.py (hunk 1 above)
Comment (dosoe) [reply beetsplug/parentwork.py]: I don't like using 'ancestor', simply because 'ancestorwork' looks bad, I prefer 'parentwork'. If I clarify it in the docs, would it be fine?

File: beetsplug/parentwork.py (hunk 3 above)
Comment (sampsyo) [reply beetsplug/parentwork.py]: Nope, they're just different variable names. If you move this into the `commands()` function, you can just call it `command` instead of `self._command` (i.e., a local variable instead of a field).

File: beetsplug/parentwork.py (hunk 1 above)
Comment (dosoe) [reply beetsplug/parentwork.py]: Ok I clarified the docstring, this should be ready to merge now.

**Comment by dosoe** [comment]
I don't understand. How can my PR change the test coverage of fetchart?

File: docs/plugins/parentwork.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+To fill in the parentwork tag and the associated parent** tags, in case there 
+are several works on the recording, it fills it with the results of the first 
+work and then appends the results of the second work only if they differ from 
+the ones already there. This is to care for cases of, for example, an opera 
+recording that contains several scenes of the opera: neither the parentwork 
+nor all the associated tags will be duplicated. 
+If there are several works linked to a recording, they all get a 
+disambiguation (empty as default) and if all disambiguations are empty, the 
+disambiguation field is left empty, else the disambiguation field can look 
+like ``,disambig,,`` (if there are four works and only the second has a
```
[the remark this answers is not in the document]
Comment (dosoe) [reply docs/plugins/parentwork.rst]: yes you're right. It's fixed now.

**Comment by sampsyo** [comment]
Arg, @codecov-io's output is pretty erratic and sometimes misleading. And we have configured it _not_ to post these pull request messages, but sometimes it just does it anyway, for some reason…

**Comment by dosoe** [comment]
I have one more question: Works on MB change from time to time, and if the work changes, I would like the parentwork to be updated as well. A change in work happens only at a ```beet mbsync``` or maybe at a ```beet up``` if ```mb_workid``` gets implemented into ```mediafile```. Right now, it doesn't get updated, so the only solution is to run ```beet parentwork -f ``` again (with the ```--force``` option else it will only look for recordings with no parentwork) and this takes a huge time because it will recover all parentworks again (which is very slow). Could there be a way to apply ```beet parentwork --force``` only to items where the ```mb_workid``` got modified? Either by applying it only to recordings that got modified in the latest time (is there a way to list those? And I mean a modification in the database, not in the file itself like ```mtime```) or by a more in-depth harmonization between mbsync, up and parentwork.

**Comment by dosoe** [comment]
One more thing, are there comments on the logging? It feels rather verbose right now, are you fine with that?

**Comment by sampsyo** [comment]
Interesting! Here's one thing I'm uncertain about: which of these two scenarios are you most concerned about adapting to?

1. The user changes the track ID for an item. So you need to get a new work to match that.
2. The user does nothing, but the MusicBrainz database changes to associate a new work with an existing track. The user wants to synchronize these new changes with the file.

The thing is that "1" above seems less likely to be an issue (how often do users _really_ change the MBID for tracks they already have in their library?), but it's easier to deal with—there are a few ways we can consider to track when the MBID has gone out of sync with the work. "2" seems like much more of a real issue, but there's essentially nothing to be done about it—to know whether anything changed, you have to check with the MB server again, which essentially means doing `beet parentwork -f`. There is basically no way around this.

---

About logging: it's probably OK to be somewhat verbose in "debug" mode. But of course, brevity is the soul of wit. I'll take a look at the log messages now.

File: beetsplug/parentwork.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+
+            parentwork_info['parent_composer'] = u', '.join(parent_composer)
+            parentwork_info['parent_composer_sort'] = u', '.join(
+                    parent_composer_sort)
+
+        if not composer_exists:
+            self._log.info(item.artist + ' - ' + item.title)
+            self._log.debug(
+                "no composer, add one at https://musicbrainz.org/work/" +
+                work_info['work']['id'])
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: It's odd that one of these lines goes to info and the other goes to debug. Here's a suggestion:

```py
self._log.debug(
    'no composer for {} - {}; add one at https://musicbrainz.org/work/{}',
    item.artist, item.title, work_info['work']['id'],
)
```

(Suitably wrapped, of course.)

File: beetsplug/parentwork.py [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        if hasattr(item, 'parentwork'):
+            hasparent = True
+        else:
+            hasparent = False
+        if not item.mb_workid:
+            self._log.info("No work attached, recording id: " +
+                           item.mb_trackid)
+            self._log.info(item.artist + ' - ' + item.title)
+            self._log.info("add one at https://musicbrainz.org" +
+                           "/recording/" + item.mb_trackid)
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: Collapsing this into one line, as above, might be nice. Especially since this goes to `info`.

File: beetsplug/parentwork.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+            return
+        if force or (not hasparent):
+            try:
+                work_info, work_date = find_parentwork_info(item.mb_workid)
+            except musicbrainzngs.musicbrainz.WebServiceError:
+                self._log.debug("Work unreachable")
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: I suggest including some evidence about the error:

```py
except musicbrainzngs.musicbrainz.WebServiceError as e:
    self._log.debug("error fetching work: {}", e)
```

File: beetsplug/parentwork.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+                return
+            parent_info = self.get_info(item, work_info)
+
+        elif hasparent:
+            self._log.debug("Work already in library, not necessary fetching")
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: Shorter wording: "work present; skipping". Including the artist/title here might be good too.

File: beetsplug/parentwork.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+            return
+
+        self._log.debug("Finished searching work for: " +
+                        item.artist + ' - ' + item.title)
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: I recommend deleting this—it seems redundant with all the other messages that explain what happened with a given item.

File: beetsplug/parentwork.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        if hasattr(item, 'parentwork'):
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: It's not clear to me why this is the right condition to switch between these two messages. They both seem to print out the parent work.

**Comment by arcresu** [comment]
Regarding the issue of updating the parent work when things change, this is an instance of the more general issue with metadata fetching covered in #266. We're planning to change how plugins work behind the scenes, and hopefully as part of that process we can solve that issue in a more satisfying way. That's a bigger issue that will take some time to get right though.

For example we could model `mbsync` as a cache of remote metadata, so it would be possible to sync just the least-recently updated items rather than the entire library. Derived metadata like the parent work could be "invalidated" if the dependent metadata (work ID) changes, triggering an immediate recalculation. This is just brainstorming rather than a concrete proposal.

**Comment by sampsyo** [comment]
Another thought occurred to me about how to simplify the logging: you don't need to do this:

```py
self._log.info('processing {} - {}', item.artist, item.title)
```

Instead, you can just do this:

```py
self._log.info('processing {}', item)
```

and the `item` will get formatted in a sensible way automatically.

File: test/test_parentwork.py [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        command_output.return_value = u'32c8943f-1b27-3a23-8660-4567f4847c94'
+        self.run_command('parentwork')
+
+        item.load()
+        self.assertEqual(item['mb_parentworkid'], u'XXX')
+
+    # test different cases, still with Matthew Passion Ouverture or Mozart
+    # requiem
+
+    def test_father_work(self, command_output):
```
Comment (arcresu) [inline test/test_parentwork.py]: Looks like you missed this test when changing the field name.
```suggestion
    def test_direct_parent_work(self, command_output):
```

File: beetsplug/parentwork.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+            if artist['type'] == 'composer':
+                if 'end' in artist.keys():
+                    work_date = artist['end']
+
+    if 'work-relation-list' in work_info['work']:
+        for work_father in work_info['work']['work-relation-list']:
+            if work_father['type'] == 'parts' \
+                    and work_father.get('direction') == 'backward':
+                father_id = work_father['work']['id']
+                return father_id, work_date
```
Comment (arcresu) [inline beetsplug/parentwork.py]: There are still some references to `father` here too.

File: docs/plugins/parentwork.rst [hunk 20]
```diff
@@ -0,0 +1,46 @@
+Parentwork Plugin
+=================
+
+The ``parentwork`` plugin fetches the work title, parentwork title and 
+parentwork composer. 
+
+In the MusicBrainz database, a recording can be associated with a work. A 
+work can itself be associated with another work, for example one being part 
+of the other (what I call the father work). This plugin looks the work id 
+from the library and then looks up the father, then the father of the father
```
Comment (arcresu) [inline docs/plugins/parentwork.rst]:
```suggestion
from the library and then looks up the direct parent, then _its_ direct parent
```

File: beetsplug/parentwork.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+
+from beets import ui
+from beets.plugins import BeetsPlugin
+
+import musicbrainzngs
+
+
+def direct_parent_id(mb_workid, work_date=None):
+    """ Given a mb_workid, find the id one of the works the work is part of
+    and the first composition date it encounters."""
```
Comment (arcresu) [inline beetsplug/parentwork.py]: For consistency with the beets coding style the final `"""` should go on its own line. There are a few places where this applies.

File: docs/plugins/parentwork.rst [hunk 22]
```diff
@@ -0,0 +1,46 @@
+Parentwork Plugin
+=================
+
+The ``parentwork`` plugin fetches the work title, parentwork title and 
+parentwork composer.
```
Comment (arcresu) [inline docs/plugins/parentwork.rst]: I know the plugin is spelled `parentwork`, but maybe we can use "parent work" when discussing the concept?

File: docs/plugins/parentwork.rst [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+be, in this case, the whole symphony. 
+
+This plugin adds five tags: 
+
+- **parentwork**: The title of the parentwork.  
+- **mb_parentworkid**: The musicbrainz id of the parentwork. 
+- **parentwork_disambig**: The disambiguation of the parentwork title. 
+- **parent_composer**: The composer of the parentwork. 
+- **parent_composer_sort**: The sort name of the parentwork composer. 
+- **work_date**: THe composition date of the work, or the first parent work
```
Comment (arcresu) [inline docs/plugins/parentwork.rst]:
```suggestion
- **work_date**: The composition date of the work, or the first parent work 
```

File: docs/plugins/parentwork.rst [hunk 24]
```diff
@@ -0,0 +1,46 @@
+Parentwork Plugin
+=================
+
+The ``parentwork`` plugin fetches the work title, parentwork title and 
+parentwork composer. 
+
+In the MusicBrainz database, a recording can be associated with a work. A 
+work can itself be associated with another work, for example one being part 
+of the other (what I call the father work). This plugin looks the work id
```
Comment (arcresu) [inline docs/plugins/parentwork.rst]:
```suggestion
of the other (what I call the direct parent work). This plugin looks the work ID 
```

**Comment by dosoe** [comment]
Thanks for the detailed comments! I think I adressed them all. 
Concerning the metadata change, what I would like to have is that ```parentwork``` notices when ```mbsync``` (or anyone else) modifies the ```mb_workid``` tag and kicks in to update the parentwork tag (and all associated tags). 
While the ```mb_trackid``` almost never changes once in the library, the ```mb_workid``` can change (and I notice it changing from time to time), so what I'm looking for is a solution to 1. Is there already something implemented that could make it easier (maybe using ```register_listener``` from the ```BeetsPlugin``` class)?

**Comment by dosoe** [comment]
I'm trying to read through https://beets.readthedocs.io/en/v1.4.9/dev/plugins.html#listen-for-events but I don't fully get it: I tried something like 
```
class ParentWorkPlugin(BeetsPlugin):
    def __init__(self):
        ...
        if self.config['auto']:
            self.register_listener('trackinfo_received', self.find_work2)
    def find_work2(self, item):
        self.find_work(self, item, True)
```
This should call ```find_work``` each time we get data from MB. However, maybe the event that I should listen to is ```database_change``` and not ```trackinfo_received``` , as the second one would trigger even if there is no change. For the first one, the parameters are ```lib``` and ```model``` and I don't know how to get an ```item``` out of them to feed to ```find_work``` . 
Additionally, as coded right now (with ```trackinfo_received```) a ```beet mbsync``` fails: 
```
Traceback (most recent call last):
  File "/usr/local/bin/beet", line 11, in <module>
    load_entry_point('beets==1.5.0', 'console_scripts', 'beet')()
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/__init__.py", line 1266, in main
    _raw_main(args)
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/__init__.py", line 1253, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/mbsync.py", line 75, in func
    self.singletons(lib, query, move, pretend, write)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/mbsync.py", line 96, in singletons
    track_info = hooks.track_for_mbid(item.mb_trackid)
  File "/usr/local/lib/python2.7/dist-packages/beets/autotag/hooks.py", line 574, in track_for_mbid
    plugins.send(u'trackinfo_received', info=track)
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 489, in send
    result = handler(**arguments)
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 149, in wrapper
    return func(*args, **kwargs)
TypeError: find_work2() takes exactly 2 arguments (1 given)
```

I'm puzzled, because I wrote ```find_work2``` specifically to have only one argument.

**Comment by dosoe** [comment]
Hi! I found my error. I changed it to: 
```
class ParentWorkPlugin(BeetsPlugin):
    def __init__(self):
        ...
        if self.config['auto']:
            self.register_listener('database_change', self.find_work2)
    def find_work2(self, lib, model):
        self.find_work(model, True)
```

This however does not work entirely fine. A try with mbsync gives: 
```
Requesting MusicBrainz release 05909b27-18c5-47dc-bcd8-971a7f25ab79
primary MB release type: album
Sending event: albuminfo_received
mbsync: applying changes to Johann Sebastian Bach; Richard Galliano - Galliano | Bach
Johann Sebastian Bach - Galliano | Bach - Art of Fugue, BWV 1080: Contrapunctus 1
  composer_sort:  -> Bach, Johann Sebastian
  work:  -> Die Kunst der Fuge, BWV 1080: I. Contrapunctus 1
  mb_workid:  -> 621651bc-7282-47ad-a401-d8be5a0988bd
  work_disambig:  -> catch all for arrangements
Sending event: database_change
parentwork: no composer for Johann Sebastian Bach - Galliano | Bach - Art of Fugue, BWV 1080: Contrapunctus 1; add one at https://musicbrainz.org/work/621651bc-7282-47ad-a401-d8be5a0988bd
parentwork: Work fetched: Die Kunst der Fuge, BWV 1080: I. Contrapunctus 1 - Johann Sebastian Bach
Johann Sebastian Bach - Galliano | Bach - Art of Fugue, BWV 1080: Contrapunctus 1
  parentwork: Die Kunst der Fuge, BWV 1080: I. Contrapunctus 1
  parent_composer_sort: Bach, Johann Sebastian
  parentwork_disambig: catch all for arrangements
  parent_composer: Johann Sebastian Bach
  mb_parentworkid: 621651bc-7282-47ad-a401-d8be5a0988bd
Sending event: database_change
Traceback (most recent call last):
  File "/usr/local/bin/beet", line 11, in <module>
    load_entry_point('beets==1.5.0', 'console_scripts', 'beet')()
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/__init__.py", line 1266, in main
    _raw_main(args)
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/__init__.py", line 1253, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/mbsync.py", line 76, in func
    self.albums(lib, query, move, pretend, write)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/mbsync.py", line 178, in albums
    apply_item_changes(lib, item, move, pretend, write)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/mbsync.py", line 40, in apply_item_changes
    item.store()
  File "/usr/local/lib/python2.7/dist-packages/beets/library.py", line 348, in store
    plugins.send('database_change', lib=self._db, model=self)
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 489, in send
    result = handler(**arguments)
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 143, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/parentwork.py", line 90, in find_work2
    self.find_work(model, True)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/parentwork.py", line 208, in find_work
    item.store()
  File "/usr/local/lib/python2.7/dist-packages/beets/library.py", line 348, in store
    plugins.send('database_change', lib=self._db, model=self)
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 489, in send
    result = handler(**arguments)
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 137, in wrapper
    assert self._log.level == logging.NOTSET
AssertionError
```
Do you know what the error is?

**Comment by sampsyo** [comment]
Hi! Can we please leave this "automation" feature to a separate pull request? It would be awesome to close the book on the basic version before we start adding fancier features.

I was even going to suggest a simpler approach to this that might be a little less surprising to users. It seems little surprising that simple metadata changes with `beet modify`, which is typically a "local only" operation, could cause network requests and stuff. Instead, maybe the plugin could keep around a flexible attribute that says "here's the work ID that the current parent work is based off of." Then, when running for a second time, check whether that still matches the current work ID and re-fetch if so. This way, the heavy lifting only happens when users ask for it rather than truly automatically.

I don't exactly see the source of that error, but it might be caused by the presence of other plugins…

**Comment by dosoe** [comment]
Ok, so I added a tag ```mb_workid_current``` which is there for this purpose, do you thing the name is appropriate? 
Also, I think that the plugin calling itself can be a real problem.

**Comment by sampsyo** [comment]
Seems reasonable to me! But can we leave this enhancement to a separate pull request? (You could even base the new branch on this branch and create the PR now.)

**Comment by dosoe** [comment]
Ok then this should be it!

File: beetsplug/parentwork.py [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+                if write:
+                    item.try_write()
+        command = ui.Subcommand(
+            'parentwork',
+            help=u'Fetches parent works, composers and dates')
+
+        command.parser.add_option(
+            u'-f', u'--force', dest='force',
+            action='store_true', default=None,
+            help=u'Re-fetches all parent works')
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: Perhaps more specifically:

> re-fetch when parent work is already present

File: beetsplug/parentwork.py [hunk 26]
```diff
@@ -0,0 +1,200 @@
...
+                item[key] = value
+
+        if work_date:
+            item['work_date'] = work_date
+        ui.show_model_changes(
+            item, fields=['parentwork', 'parentwork_disambig',
+                          'mb_parentworkid', 'parent_composer',
+                          'parent_composer_sort', 'work_date'])
+
+        item.store()
```
Comment (sampsyo) [inline beetsplug/parentwork.py]: Delete this call to `item.store()`; the item is already stored in both places where this function is called.

Comment (dosoe) [reply beetsplug/parentwork.py]: this might actually solve the problem I had with catching database changes, as the error was on the ```item.store()``` . But it's for another time I guess.

**Comment by sampsyo** [comment]
Merged! Thank you for your long-term persistence in seeing this through!
